api/README.md

```diff
@@ -0,0 +1,264 @@
+
+The main difference with the k8s API's definition of compatibility here is that it is not required to be able round-trip a change (convert to different API versions and back) with no loss of information, since contrary to k8s' API we don't need to store API objects here.
+
+As long as a change abides by the compatibility rules outlined above, one can simply amend the `proto` file defining that API version, and then regenerate all the Go boilerplate code.
+
+### How to add a new version to an existing API group
+
+Any change that cannot be done in a compatible way requires creating a new API version.
+
+Steps to add a new API version:
+1. define it its own new `api.proto` file
+2. generate the Go protobuf code
+3. update the API group's internal representations (in its `types.go` file) to be able to represent all of the group's version (the new and the old ones)
+3. add any needed conversion functions for all existing versions of this API group to account for the changes made at the previous step
+4. re-generate all of the Go boilerplate code
+5. now you can change the API group's server to add your new feature!
+
+### How to deprecate, and eventually remove...
+
+From the CSI [proxy KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-windows/20190714-windows-csi-support.md#csi-proxy-grpc-api-graduation-and-deprecation-policy):
+
+> In accordance with standard Kubernetes conventions, the above API will be introduced as v1alpha1 and graduate to v1beta1 and v1 as the feature graduates. Beyond a vN release in the future, new RPCs and enhancements to parameters will be introduced through vN+1alpha1 and graduate to vN+1beta1 and vN+1 stable versions as the new APIs mature.
+>
+> Members of CSIProxyService API may be deprecated and then removed from csi-proxy.exe in a manner similar to Kubernetes deprecation [policy](https://kubernetes.io/docs/reference/using-api/deprecation-policy/) although maintainers will make an effort to ensure such deprecation is as rare as possible. After their announced deprecation, a member of CSIProxyService API must be supported:
+>
+> 1. 12 months or 3 releases (whichever is longer) if the API member is part of a Stable/vN version.
+> 2. 9 months or 3 releases (whichever is longer) if the API member is part of a Beta/vNbeta1 version.
+> 3. 0 releases if the API member is part of an Alpha/vNalpha1 version.
+> To continue running CSI Node Plugins that depend on an old version of csi-proxy.exe, vN, some of whose members have been removed, Kubernetes administrators will be required to run the latest version of the csi-proxy.exe (that will be used by CSI Node Plugins that use versions of CSIProxyService more recent than vN) along with an old version of csi-proxy.exe that does support vN.
+
+With that in mind, each subsection below details how to deprecate, then remove:
+
+#### A field in an API object
+
+Mark it as deprecated in the `proto` file, e.g.:
```

Comment:
To clarify, all of these types of deprecation results in deprecating the entire api version right? Because in order to actual remove a field or proc, you need to serve a new api version, and remove the current one.

Reply:
I don't think that deprecating a single field, or a single procedure, necessarily means deprecating the entire API version. There's no strong reason to deprecate API versions as fast as possible, and adding a new one shouldn't necessarily mean deprecating all the previous versions for that group. And deprecation doesn't mean immediate removal either.

Reply:
I guess what I'm wondering about is to actually remove support for a field or function entirely, we would have to deprecate and eventually remove the entire API version (and replace it with a new api version that doesn't have it)

Reply:
Absolutely; deprecating a single field or procedure is simply meant to let users know that this field or procedure will be removed in a subsequent version. It doesn't change anything to the behaviour of the version it's applied to.